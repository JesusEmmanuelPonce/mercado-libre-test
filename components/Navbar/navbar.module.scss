.navbar {
    display: flex;
    justify-content: center;

    background: #ffe600;
    padding: 10px;
    width: 100%;

    &__search {
        align-items: center;
        display: flex;
        gap: 25px;

        max-width: 1024px;
        width: 100%;

        &__btn {
            background-color: transparent;
            border: transparent;
            cursor: pointer;
            padding: 0;
        }
    }
}
