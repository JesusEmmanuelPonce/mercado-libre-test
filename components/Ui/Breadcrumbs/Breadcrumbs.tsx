import { FC } from "react";

import styles from "./breadcrumbs.module.scss";

interface IBreadcrumbsProps {};

const Breadcrumbs: FC<IBreadcrumbsProps> = ({}) => {
    return (
        <ul className={styles.breadcrumbWrap}>
            <li>Electronica, Audio y Video {" > "}</li>
            <li>iPod {" > "}</li>
            <li>Reproductores {" > "}</li>
            <li>iPod touch {" > "}</li>
            <li className={styles.active}>32 GB</li>
        </ul>
    )
};

export default Breadcrumbs;
