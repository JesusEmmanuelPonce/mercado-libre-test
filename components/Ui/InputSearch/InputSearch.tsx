import { useRouter } from 'next/router';
import { ChangeEvent, FC, KeyboardEvent, useState } from "react";

import Image from "next/image";
import styles from "./inputSearch.module.scss"
import Ic_Search from "/public/img/ic_Search.png"

interface IInputSearchProps {};

const InputSearch: FC<IInputSearchProps> = ({}) => {

    const router = useRouter();

    const [search, setSearch] = useState<string>("")

    const handleSearch = () => {
        router.push(`/search?q=${search}`)
    };

    const onKeyPressSearch = ({ code }: KeyboardEvent<HTMLInputElement>) => {
        if(code === "Enter") {
            handleSearch();
        }
    };

    return (
        <div className={styles.inputSearch}>
            <input
                type="text"
                value={search}
                placeholder="Nunca dejes de buscar"
                onChange={({ target: { value } }: ChangeEvent<HTMLInputElement>) => setSearch(value)}
                onKeyPress={onKeyPressSearch}
            />
            <button
                type="button"
                className={styles.inputSearch__btn}
                onClick={handleSearch}
            >
                 <Image
                    src={Ic_Search}
                    alt="input search icon"
                    width={20}
                    height={20}
                />
            </button>
        </div>
    )
}

export default InputSearch;