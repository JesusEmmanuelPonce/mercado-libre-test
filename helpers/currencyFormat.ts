export const currencyFormat = (amount: number | string ) => {

    let amountNumber = 0;

    if(typeof amount === 'string'){
        amountNumber = parseFloat(amount);
    } else{
        amountNumber = amount
    }

    const currency = Intl.NumberFormat("en-US", {
        style: "currency",
        currency: "USD",
    });

    return currency.format(amountNumber);
};
