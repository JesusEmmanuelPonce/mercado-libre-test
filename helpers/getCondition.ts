export const getCondition = (condition: string) => {
    switch (condition) {
        case "new":
            return "Nuevo";
    
        default: 
            return condition;
    }
};
