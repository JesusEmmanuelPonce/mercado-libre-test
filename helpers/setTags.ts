export const capitalize = (str: string = "") => {
    const lower = str.toLowerCase();

    return str.charAt(0).toLocaleUpperCase() + lower.slice(1);
};

export const setTitleTag = (str: string = "") => {

    const result = `${capitalize(str)} | MercadoLibre ğŸ“¦`

    return result;
};
