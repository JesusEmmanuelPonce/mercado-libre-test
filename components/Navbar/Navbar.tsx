import Image from 'next/image';
import { FC } from 'react';
import { useRouter } from 'next/router';

import Logo  from "/public/img/logo.png"
import InputSearch from '../Ui/InputSearch';
import styles from "./navbar.module.scss";

interface INavbarProps {};

const Navbar: FC<INavbarProps> = ({}) => {

    const router = useRouter();

    const handleToHome = () => {
        router.push("/");
    };

    return (
        <nav className={styles.navbar}>
            <section className={styles.navbar__search}>
                <button
                    type='button'
                    onClick={handleToHome}
                    className={styles.navbar__search__btn}
                >
                    <Image
                        src={Logo}
                        alt="input search icon"
                        height={40}
                        width={60}
                        layout="fixed"
                    />
                </button>
                <InputSearch />
            </section>
        </nav>
    )
};

export default Navbar;
