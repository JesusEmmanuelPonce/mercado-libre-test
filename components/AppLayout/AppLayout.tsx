import { FC, ReactNode } from 'react';

import Navbar from '../Navbar';
import Breadcrumbs from '../Ui/Breadcrumbs';
import styles from "./applayout.module.scss";

interface IAppLayoutProps {
    children: ReactNode;
}

const AppLayout: FC<IAppLayoutProps> = ({ children }) => {
    return (
        <>
            <Navbar />
            <main className={styles.main}>
			    <Breadcrumbs />
                {children}
            </main>
        </>
    )
}

export default AppLayout;