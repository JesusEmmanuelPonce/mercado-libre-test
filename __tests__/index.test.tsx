import fetch from 'jest-fetch-mock';
import { create } from "react-test-renderer";
import { render } from '@testing-library/react';
import '@testing-library/jest-dom';

import Home from '../pages';
import Navbar from '../components/Navbar';
import AppLayout from '../components/AppLayout';
import Breadcrumbs from '../components/Ui/Breadcrumbs';
import InputSearch from "../components/Ui/InputSearch";
import axiosClient from "../helpers/axiosClient";
import ProductDetail from "../pages/items/[id]";
import { getCondition } from '../helpers/getCondition';
import { currencyFormat } from '../helpers/currencyFormat';
import { capitalize, setTitleTag } from '../helpers/setTags';

jest.mock('next/router', () => ({
	useRouter() {
		return {
			router: "/",
			pathname: "",
			query: "",
			asPath: "/",
			push: jest.fn()
		}
	}
}));

jest.spyOn(require("next/router"), "useRouter")

global.fetch = require('jest-fetch-mock');

// Services
describe("Fetch Api to all products", () => {
	beforeEach(() => {
		fetch.resetMocks()
	});

	test("Call to All products", async() => {

		const { data } = await axiosClient.get("https://api.mercadolibre.com/sites/MLA/search?q=producto&limit=4")

		expect(data.results[0].id).toMatch("MLA")
	});
});

describe("Fetch Api to product detail", () => {
	beforeEach(() => {
		fetch.resetMocks()
	});

	test("Call to product detail", async() => {

		const { data } = await axiosClient.get("https://api.mercadolibre.com/items/MLA884180079")

		expect(data.id).toMatch(/MLA/)
	});
});

describe("Fetch Api to product description", () => {
	beforeEach(() => {
		fetch.resetMocks()
	});

	test("Call to product detail", async() => {

		const { data } = await axiosClient.get("https://api.mercadolibre.com/items/MLA884180079/description")

		expect(data.snapshot.url).toMatch(/descriptions.mlstatic.com/)
	});
});

// Helpers
describe("Currency format helper", () => {
	const currency = currencyFormat(100);

	test("return number formatted", () => {
		expect(currency).toBe("$100.00");
	});
});

describe("Capitalize helper", () => {
	const sentence = capitalize("hello world");

	test("Capitalize a sentence", () => {
		expect(sentence).toBe("Hello world");
	});
});

describe("Set title tag helper", () => {
	const sentence = setTitleTag("xbox");

	test("Capitalize params and join params with the sentence | Mercado libre", () => {
		expect(sentence).toBe("Xbox | MercadoLibre ğŸ“¦");
	});
});

describe("Get conditions helper", () => {
	const word = getCondition("new");

	test("Change lang to spanish, if the word doesnt register return same", () => {
		expect(word).toBe("Nuevo");
	});
});

// Components
describe("Breadcrumbs snapshot", () => {

	test('render breadcrumbs', () => {
		render(<Breadcrumbs />);
	});

	test('Check footer component UI', () => {
		const breadcrumbs = create(<Breadcrumbs />)
		expect(breadcrumbs.toJSON()).toMatchSnapshot()
	});
});

describe("Input search snapshot", () => {

	test('render input search', () => {
		render(<InputSearch />);
	});

	test('Check input search component UI', () => {
		const inputsearch = create(<InputSearch />)
		expect(inputsearch.toJSON()).toMatchSnapshot()
	});
});

describe("Navbar component", () => {

	test('render navbar component', () => {
		render(<Navbar />)
  	});
});

describe("App layout component", () => {

	test('render app layout component', () => {
		render(<AppLayout />)
  	});

    test("App layout title rendering", () => {
		const text = "Mercado libre";

        render(<AppLayout titleTag={text} />);

        expect(text).toMatch(/Mercado libre/);
    });
});

// Pages
describe("Search page", () => {

	test('Render search page', () => {
		render(<Home />)
  	});
});

describe("Product detail page", () => {
	const product = {
		author: {
			name: "Jesus",
			lastname: "Ponce"
		},
		item: {
			id: "MLA884180079",
			title: "Product",
			price: {
				currency: "ARS",
				amount: 12,
				decimals: 12,
			},
			picture: "thumbnail",
			condition: "new",
			free_shipping: true,
			sold_quantity: 2,
			description: "description",
		}
	}

	test('Render product detail page', () => {
		render(<ProductDetail product={product}/>)
  	});
});

