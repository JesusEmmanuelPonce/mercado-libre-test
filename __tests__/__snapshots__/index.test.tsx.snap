// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Breadcrumbs snapshot Check footer component UI 1`] = `
<ul
  className="breadcrumbWrap"
>
  <li>
    Electronica, Audio y Video 
     &gt; 
  </li>
  <li>
    iPod 
     &gt; 
  </li>
  <li>
    Reproductores 
     &gt; 
  </li>
  <li>
    iPod touch 
     &gt; 
  </li>
  <li
    className="active"
  >
    32 GB
  </li>
</ul>
`;

exports[`Input search snapshot Check input search component UI 1`] = `
<div
  className="inputSearch"
>
  <input
    onChange={[Function]}
    onKeyPress={[Function]}
    placeholder="Nunca dejes de buscar"
    type="text"
    value=""
  />
  <button
    className="inputSearch__btn"
    onClick={[Function]}
    type="button"
  >
    <span
      style={
        {
          "background": "none",
          "border": 0,
          "boxSizing": "border-box",
          "display": "inline-block",
          "height": 20,
          "margin": 0,
          "opacity": 1,
          "overflow": "hidden",
          "padding": 0,
          "position": "relative",
          "width": 20,
        }
      }
    >
      <img
        alt="input search icon"
        data-nimg="fixed"
        decoding="async"
        onError={[Function]}
        onLoad={[Function]}
        src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
        style={
          {
            "border": "none",
            "bottom": 0,
            "boxSizing": "border-box",
            "display": "block",
            "height": 0,
            "left": 0,
            "margin": "auto",
            "maxHeight": "100%",
            "maxWidth": "100%",
            "minHeight": "100%",
            "minWidth": "100%",
            "objectFit": undefined,
            "objectPosition": undefined,
            "padding": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
            "width": 0,
          }
        }
      />
      <noscript>
        <img
          alt="input search icon"
          data-nimg="fixed"
          decoding="async"
          loading="lazy"
          src="/_next/image?url=%2Fimg.jpg&w=48&q=75"
          srcSet="/_next/image?url=%2Fimg.jpg&w=32&q=75 1x, /_next/image?url=%2Fimg.jpg&w=48&q=75 2x"
          style={
            {
              "border": "none",
              "bottom": 0,
              "boxSizing": "border-box",
              "display": "block",
              "height": 0,
              "left": 0,
              "margin": "auto",
              "maxHeight": "100%",
              "maxWidth": "100%",
              "minHeight": "100%",
              "minWidth": "100%",
              "objectFit": undefined,
              "objectPosition": undefined,
              "padding": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
              "width": 0,
            }
          }
        />
      </noscript>
    </span>
  </button>
</div>
`;
